coderd/provisionerdserver/provisionerdserver_test.go

```diff
@@ -1804,6 +1804,155 @@ func TestNotifications(t *testing.T) {
+			},
+			{
+				name:           "AutoBuild",
+				tplOwner:       ownerA,
+				buildInitiator: ownerA,
+				reason:         database.BuildReasonAutostart,
+				transition:     database.WorkspaceTransitionStart,
+				// Ensure that during automated builds, all template admins and owners
+				// receive notifications.
+				receivers: []uuid.UUID{ownerA.ID, ownerB.ID, tplAdmin.ID},
+			},
+		}
+
+		for _, c := range tc {
+			t.Run(c.name, func(t *testing.T) {
+				// Given: a template and a workspace build
+				isManualBuild := c.reason == database.BuildReasonInitiator
+				template := dbgen.Template(t, db, database.Template{
+					Name:           "template",
+					Provisioner:    database.ProvisionerTypeEcho,
+					OrganizationID: pd.OrganizationID,
+					CreatedBy:      c.tplOwner.ID,
+				})
+				template, err := db.GetTemplateByID(ctx, template.ID)
+				require.NoError(t, err)
+				version := dbgen.TemplateVersion(t, db, database.TemplateVersion{
+					OrganizationID: pd.OrganizationID,
+					TemplateID: uuid.NullUUID{
+						UUID:  template.ID,
+						Valid: true,
+					},
+					JobID: uuid.New(),
+				})
+				workspaceOwner := member
+				workspace := dbgen.Workspace(t, db, database.Workspace{
+					TemplateID:     template.ID,
+					OwnerID:        workspaceOwner.ID,
+					OrganizationID: pd.OrganizationID,
+				})
+				b := database.WorkspaceBuild{
+					WorkspaceID:       workspace.ID,
+					TemplateVersionID: version.ID,
+					Transition:        c.transition,
+					Reason:            c.reason,
+				}
+				// Set the build initiator if the test case specifies one for manual builds.
+				if isManualBuild {
+					b.InitiatorID = c.buildInitiator.ID
+				}
+				build := dbgen.WorkspaceBuild(t, db, b)
+				file := dbgen.File(t, db, database.File{CreatedBy: c.buildInitiator.ID})
+				job := dbgen.ProvisionerJob(t, db, ps, database.ProvisionerJob{
+					FileID: file.ID,
+					Type:   database.ProvisionerJobTypeWorkspaceBuild,
+					Input: must(json.Marshal(provisionerdserver.WorkspaceProvisionJob{
+						WorkspaceBuildID: build.ID,
+					})),
+					OrganizationID: pd.OrganizationID,
+				})
+				_, err = db.AcquireProvisionerJob(ctx, database.AcquireProvisionerJobParams{
+					OrganizationID: pd.OrganizationID,
+					WorkerID: uuid.NullUUID{
+						UUID:  pd.ID,
+						Valid: true,
+					},
+					Types: []database.ProvisionerType{database.ProvisionerTypeEcho},
+				})
+				require.NoError(t, err)
+
+				// When: the build fails
+				_, err = srv.FailJob(ctx, &proto.FailedJob{
+					JobId: job.ID.String(),
+					Type: &proto.FailedJob_WorkspaceBuild_{
+						WorkspaceBuild: &proto.FailedJob_WorkspaceBuild{
+							State: []byte{},
+						},
+					},
+				})
+				require.NoError(t, err)
+
+				// Then: send the template build failed notifications
+				var buildFailedNotifications []*testutil.Notification
+				for _, n := range notifEnq.Sent {
+					if n.TemplateID == notifications.TemplateTemplateBuildFailed {
```

Comment:
What if they're not the template you're expecting?

Your tests should not only prove that something works; it should also prove that it _only_ works in _precisely_ the way you're expecting, and any relevant deviation should fail the test.